thriftbp/ttl_client.go

```diff
@@ -49,18 +79,70 @@ func (c *ttlClient) IsOpen() bool {
 	return true
 }
 
-// newTTLClient creates a ttlClient with a thrift TTransport and a ttl+jitter.
-func newTTLClient(transport thrift.TTransport, client thrift.TClient, ttl time.Duration, jitter float64) *ttlClient {
-	var expiration time.Time
+// refresh is called when the ttl hits to try to refresh the connection.
+func (c *ttlClient) refresh() {
+	client, transport, err := c.generator()
+	if err != nil {
+		// We cannot replace this connection in the background,
+		// leave client and transport be,
+		// this connection will be replaced by the pool upon next use.
+		c.replaceCounter.With("success", "False").Add(1)
+		return
+	}
+
+	// replace with the refreshed connection
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.closed {
+		// If Close was called after we entered this function,
+		// close the newly created connection and return early.
+		transport.Close()
+		return
+	}
+	c.renew(time.Now())
+	c.client = client
+	if c.transport != nil {
+		// close the old transport before replacing it, to avoid connection leaks.
+		c.transport.Close()
+	}
+	c.transport = transport
+	c.replaceCounter.With("success", "True").Add(1)
+}
+
+// renew updates expiration and timer base on the given timestamp.
+//
+// caller of this function MUST already held the lock.
```

Comment:
🔕 🔕 (very optional)

This "must hold the lock" is when I pivot to using a 1-buffered channel as a mutex with all of the mutex-guarded state inside it.

Then you attach "must hold the lock" methods on that state struct.  Now it is much harder to call it wrong.

Reply:
do you have an example? I know how to use 1-buffer channel to replace a lock, but I'm not sure how to use it in a function that you are supposed to be already holding the lock.

Reply:
I guess https://github.com/reddit/baseplate.go/pull/438/commits/0cecf4d4f707b426e630011e802159b5e1047e3a is what you meant?